File: android-java/app/src/main/java/com/example/dittotasks/Task.java
package com.example.dittotasks;

import com.ditto.kotlin.DittoQueryResultItem;

import org.json.JSONException;
import org.json.JSONObject;

public class Task {
    private final String id;
    private final String title;
    private final boolean done;
    private final boolean deleted;

    public Task(String title) {
        this(null, title, false, false);
    }

    public Task(String id, String title, boolean done, boolean deleted) {
        this.id = id;
        this.title = title;
        this.done = done;
        this.deleted = deleted;
    }

    public static Task fromQueryItem(DittoQueryResultItem item) {
        try {
            JSONObject json = new JSONObject(item.jsonString());
            return new Task(
                    json.isNull("_id") ? null : json.optString("_id", null),
                    json.isNull("title") ? null : json.optString("title", null),
                    json.optBoolean("done", false),
                    json.optBoolean("deleted", false));
        } catch (JSONException e) {
            throw new RuntimeException("Failed to parse task from query result", e);
        }
    }

    public String getId() {
        return id;
    }

    public String getTitle() {
        return title;
    }

    public boolean isDone() {
        return done;
    }

    public boolean isDeleted() {
        return deleted;
    }
}

File: android-kotlin/QuickStartTasks/app/src/main/java/live/ditto/quickstart/tasks/data/Task.kt
package live.ditto.quickstart.tasks.data

import android.util.Log
import org.json.JSONException
import org.json.JSONObject
import java.util.UUID

data class Task(
    val _id: String = UUID.randomUUID().toString(),
    val title: String,
    val done: Boolean = false,
    val deleted: Boolean = false,
) {
    fun toMap(): Map<String, Any?> = mapOf(
        "_id" to _id,
        "title" to title,
        "done" to done,
        "deleted" to deleted
    )

    companion object {
        private const val TAG = "Task"

        fun fromJson(jsonString: String): Task {
            return try {
                val json = JSONObject(jsonString)
                Task(
                    _id = json.optString("_id", UUID.randomUUID().toString()),
                    title = json.optString("title", ""),
                    done = json.optBoolean("done", false),
                    deleted = json.optBoolean("deleted", false)
                )
            } catch (e: JSONException) {
                Log.e(TAG, "Unable to convert JSON to Task", e)
                Task(title = "")
            }
        }
    }
}
